relay-server/src/services/upload.rs

```
@@ -0,0 +1,202 @@
+    inflight: FuturesUnordered<BoxFuture<'static, UploadResult>>,
+    max_inflight: usize,
+    timeout: std::time::Duration,
+}
+
+impl UploadService {
+    pub fn new(config: &UploadServiceConfig) -> Self {
+        let UploadServiceConfig {
+            max_inflight,
+            timeout,
+        } = config;
+        Self {
+            inflight: FuturesUnordered::new(),
+            max_inflight: *max_inflight,
+            timeout: *timeout,
+        }
+    }
+}
+
+impl Service for UploadService {
+    type Interface = Upload;
+
+    async fn run(mut self, mut rx: Receiver<Self::Interface>) {
+        loop {
+            tokio::select! {
+                biased;
+
+                Some(result) = self.inflight.next() => {
+                    relay_statsd::metric!(gauge(RelayGauges::AttachmentUploadsInFlight) = self.inflight.len() as u64);
+                    if let Some((uploaded_attachment, respond_to)) = result {
+                        respond_to.send(uploaded_attachment);
+                    }
+                }
+                Some(message) = rx.recv() => {
+                    let Upload { attachment, respond_to } = message;
+                    if self.inflight.len() > self.max_inflight {
+                        // Load shed to prevent backlogging in the service queue and affecting other parts of Relay.
+                        // We might want to have a less aggressive mechanism in the future.
+                        drop(attachment.reject_err(Error::LoadShed));
+                        continue;
+                    }
+
+                    self.inflight.push(managed_upload(self.timeout, attachment, respond_to).boxed());
+                    relay_statsd::metric!(gauge(RelayGauges::AttachmentUploadsInFlight) = self.inflight.len() as u64);
+                }
+            }
+        }
+    }
+}
+
+type UploadResult = Option<(
+    Managed<UploadedAttachment>,
+    Recipient<Managed<UploadedAttachment>, NoResponse>,
+)>;
```

It looks like in all `None` branches you call `managed.reject` and discard the result, this could instead be `Result<_, Rejected<()>>` making sure an outcome is always emitted.

Possibly `Rejected<E>` if you want to propagate the error.

---

This is roughly what I had in a first iteration, but then I thought what's the point of returning an error if I'm not using it? Just to signal to the caller that outcomes have been emitted?

---

A little bit of type safety that outcomes have been dealt with, e.g. if you refactor you can't accidentally forget to emit outcomes.

---

Refactored to return `Rejected` now, that's also nicer for emitting metrics.